File: iOSDemoProject/iOSDemoProject/Router/AppRouter.swift
import UIKit

final class AppRouter {
    static let shared = AppRouter()

    private weak var navigationController: UINavigationController?

    private init() {}

    func setup(with navigationController: UINavigationController) {
        self.navigationController = navigationController
    }

    func navigate(to urlString: String, from viewController: UIViewController? = nil) {
        guard let url = URL(string: urlString), let host = url.host else { return }

        let path = url.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let params = url.queryParameters

        let targetVC: UIViewController

        switch host {
        case "native":
            targetVC = resolveNative(path: path, params: params)
        case "rn":
            targetVC = RNContainerViewController(moduleName: path.isEmpty ? "home" : path, initialProperties: params)
        case "flutter":
            targetVC = FlutterContainerViewController(route: path.isEmpty ? "/" : "/\(path)", params: params)
        case "webview":
            let webURL = params["url"] ?? ""
            targetVC = WebViewContainerViewController(urlString: webURL)
        default:
            return
        }

        let nav = navigationController
            ?? (viewController as? UINavigationController)
            ?? viewController?.navigationController

        DispatchQueue.main.async {
            nav?.pushViewController(targetVC, animated: true)
        }
    }

    func pop() {
        DispatchQueue.main.async {
            self.navigationController?.popViewController(animated: true)
        }
    }

    private func resolveNative(path: String, params: [String: String]) -> UIViewController {
        switch path {
        case "demo":
            return NativeDemoViewController()
        case "home":
            return ViewController()
        default:
            return ViewController()
        }
    }
}

File: AndroidDemoProject/app/src/main/java/com/by/androiddemoproject/router/AppRouter.kt
package com.by.androiddemoproject.router

import android.content.Context
import android.content.Intent
import android.net.Uri
import com.by.androiddemoproject.MainActivity
import com.by.androiddemoproject.bridges.flutter.FlutterContainerActivity
import com.by.androiddemoproject.bridges.rn.RNContainerActivity
import com.by.androiddemoproject.bridges.webview.WebViewContainerActivity
import com.by.androiddemoproject.demo.NativeDemoActivity

object AppRouter {

    fun navigate(context: Context, url: String) {
        val uri = Uri.parse(url)
        val host = uri.host ?: return
        val path = uri.path?.trimStart('/') ?: ""
        val params = uri.queryParameters()

        val intent: Intent = when (host) {
            "native" -> resolveNative(context, path, params)
            "rn" -> Intent(context, RNContainerActivity::class.java).apply {
                putExtra("moduleName", if (path.isEmpty()) "home" else path)
                putExtra("params", HashMap(params))
            }
            "flutter" -> Intent(context, FlutterContainerActivity::class.java).apply {
                putExtra("route", if (path.isEmpty()) "/" else "/$path")
                putExtra("params", HashMap(params))
            }
            "webview" -> Intent(context, WebViewContainerActivity::class.java).apply {
                putExtra("url", params["url"] ?: "")
            }
            else -> return
        }

        context.startActivity(intent)
    }

    private fun resolveNative(context: Context, path: String, params: Map<String, String>): Intent {
        return when (path) {
            "demo" -> Intent(context, NativeDemoActivity::class.java)
            "home" -> Intent(context, MainActivity::class.java)
            else -> Intent(context, MainActivity::class.java)
        }
    }
}
